### DISCUSSION TITLE
Why use system prompt to choose function instead of OpenAI function calling?

### DISCUSSION THREAD

**Discussion by lyzhang1999** [open-webui/pipelines#223] [Ideas]
blueprints/function_calling_blueprint.py:
```
        # System prompt for function calling
        fc_system_prompt = (
            f"Tools: {json.dumps(tools_specs, indent=2)}"
            + """
If a function tool doesn't match the query, return an empty string. Else, pick a function tool, fill in the parameters from the function tool's schema, and return it in the format { "name": \"functionName\", "parameters": { "key": "value" } }. Only pick a function if the user asks.  Only return the object. Do not return any other text."
"""
        )
```

Use system prompt for function calling is not native usage, Why not use openai function calling like:

```
tools = [
    {
        "type": "function",
        "function": {
            "name": "get_delivery_date",
            "description": "Get the delivery date for a customer's order. Call this whenever you need to know the delivery date, for example when a customer asks 'Where is my package'",
            "parameters": {
                "type": "object",
                "properties": {
                    "order_id": {
                        "type": "string",
                        "description": "The customer's order ID.",
                    },
                },
                "required": ["order_id"],
                "additionalProperties": False,
            },
        }
    }
]

messages = [
    {"role": "system", "content": "You are a helpful customer support assistant. Use the supplied tools to assist the user."},
    {"role": "user", "content": "Hi, can you tell me the delivery date for my order?"}
]

response = openai.chat.completions.create(
    model="gpt-4o",
    messages=messages,
    tools=tools,
)
```

**Comment by hasani114** [comment]
The function calling is somewhat broken because of this in my experience. If I enable multiple functions it seems to go crazy and it seems like it might be because of this.

My initial thought was it was done so that models that don't support openai function spec can still work but then I realized the whole idea behind open webui is "openai compatible api"..

**Reply by justinh-rahb** [reply]
_There wasn't a spec_ when we started down this road, but now OpenAI and Ollama seem to be heading in a similar direction. You're correct that ours was done this way for compatibility reasons: it works with any backend as long as the model is pretrained for function calling, there's no explicit support for it required from the API endpoint this way.

**Reply by hasani114** [reply]
That makes sense. Is there a way to edit the system message / template for the function decider LLM call? I can't seem to find it in Admin Settings. I see the template for Documents and Title Generation / Web Search but not for Function/Tool Calling.

Also, any way to change the model that is being used by default? Would I just edit the function_calling_blueprint directly?

I did further tests and am confused. It would be great if you can help a bit.

When I use a tool, in the open webui server logs I see this: 
![image](https://github.com/user-attachments/assets/e5b92338-a202-4529-a507-2d47f74265d1)

However, I also intercepted the request that is going to the model using a filter and the body that it prints is this:

![image](https://github.com/user-attachments/assets/e10cd11d-5c80-4408-a3a6-e11aae956d75)

If I enable multiple tools, the filter prints out the body for each tool. Is the model being called multiple times for each tool that is enabled? Even if the question just pertains to one tool?

![image](https://github.com/user-attachments/assets/3c178411-ad0c-42b4-a9df-1cbff1ed9cdf)

I thought the actual model was only getting the "tools_context" that is generated by another filter/chain that is hard coded? Can you explain the lifecycle of the actual request during the default tool usage feature?
